### PR TITLE
New Foundation (WIP)

### PR DESCRIPTION
**Pull request by efriis**
Convenience methods and tests aren't present yet

### CODE REVIEW

File: proto/entity.proto [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+message Interval {
+  required double a = 1;
+  required double b = 2;
+}
+
+message Point {
+  required double x = 1;
+  required double y = 2;
+  required int32 page_index = 3;
```
Comment (andrey-ib) [inline proto/entity.proto:11]: It seems kinda funky to have `page_index` attributes on the `Point` and `BBox` (which is really just `Rectangle`) types. The concepts of point and rectangle are independent of the concept of a page. What do you think about adding a `Rectangle` type and implementing `BBox` as `{page_index: int, rect: Rectangle}`? And something similar for a `PointInPage` type?

File: py/foundation/geometry.py [hunk 2]
```diff
@@ -95,17 +100,21 @@ def intersection(Is: Iterable['Interval']) -> Optional['Interval']:
...
   x: float
   y: float
+  page_index: int
+
+  @property
+  def type(self) -> str:
+    return "Point"
 
   def __str__(self) -> str:
     return "Point({}, {})".format(self.x, self.y)
```
Comment (andrey-ib) [inline py/foundation/geometry.py:120]: This is missing `page_index`.

File: py/foundation/geometry.py [hunk 3]
```diff
@@ -130,15 +139,19 @@ def max_x(points: Iterable['Point']) -> float:
   def max_y(points: Iterable['Point']) -> float:
```
Comment (andrey-ib) [inline py/foundation/geometry.py:139]: This and all the other geometric methods (`distance`, etc.) of `Point` are now wrong -- they all assume that all input points have the same `page_index`.

The same goes for the `BBox` member methods.

File: py/foundation/geometry.py [hunk 4]
```diff
@@ -188,22 +201,27 @@ def percentages_overlapping(self, other: 'BBox') -> Optional['BBox']:
...
+  def build(ix: Optional[Interval], iy: Optional[Interval], page_index: Optional[int]) -> Optional['BBox']:
+    return BBox(ix, iy, page_index) if ix is not None and iy is not None and page_index is not None else None
 
   @staticmethod
-  def spanning(ps: Iterable[Point]) -> Optional['BBox']:
+  def spanning(ps: Iterable[Point]) -> 'BBox':
     ps = tuple(ps)
     if not ps:
-      return None
+      raise ValueError('Cannot get spanning of an empty iterrable')
```
Comment (andrey-ib) [inline py/foundation/geometry.py:220]: Spelling error

File: proto/entity.proto [hunk 5]
```diff
@@ -0,0 +1,46 @@
+syntax = "proto2";
```
Comment (vontell) [inline proto/entity.proto:1]: I thought we aren't using proto now?

File: proto/entity.proto (hunk 1 above)
Comment (efriis) [reply proto/entity.proto:11]: Makes sense. Will change. Note that this proto file will be deleted shortly (this was just for an experiment)

File: py/foundation/geometry.py (hunk 2 above)
Comment (efriis) [reply py/foundation/geometry.py:120]: Will implement the previous comment instead

File: Makefile [hunk 6]
```diff
@@ -1,2 +1,5 @@
 test:
 	./python_tests.sh
+
+avro:
+	PYTHONPATH=$(PYTHONPATH):./py/ python scripts/generate_avro.py
```
Comment (andrey-ib) [inline Makefile:5]: If you mean `python3`, then you should say that. At least I don't have my machine configured the Instabase Way™ and I don't think Dominique does either.

**Pull request closed without merging**